src/pymor/algorithms/rand_la.py

```
@@ -2,21 +2,266 @@
+
+        with
+
+        .. math::
+            c_{\mathrm{est}}
+            =\frac{1}{\sqrt{2\lambda_{\mathrm{min}}}\operatorname{erf}^{-1}
+            \left(\left(\frac{\texttt{p_fail}}{n}\right)^{1/\texttt{num_testvecs}}\right)},
+
+        where :math:`\Omega` is a set of `num_testvecs` random vectors, :math:`S` denotes the inner
+        product of the range, :math:`T` denotes inner product of the source of :math:`A` and
+        :math:`n=\min\{\dim(S),\,\dim(T)\}`.
+
+        Parameters
+        ----------
+        basis_size
+            The size of the basis.
+        num_testvecs
+            Number of test vectors for error estimation.
+        p_fail
+            Maximum failure probabilty of the error estimate.
+
+        Returns
+        -------
+        err
+            The approximate error of the basis.
+        """
+        assert isinstance(basis_size, int) and basis_size > 0
+        if basis_size > min(self.A.source.dim, self.A.range.dim):
+            self.logger.warning('Requested basis is larger than the rank of the operator!')
+            self.logger.info('Proceeding with maximum operator rank.')
+            basis_size = min(self.A.source.dim, self.A.range.dim)
+        assert 0 < num_testvecs and isinstance(num_testvecs, int)
+        assert 0 < p_fail < 1
+
+        err = self._estimate_error(basis_size, num_testvecs, p_fail)
+        self.logger.info(f'Estimated error (basis dimension {basis_size}): {err:.5e}.')
+        return err
+
+    def _extend_basis(self, k):
+        W = self.A.source.random(k, distribution='normal')
+        if self.complex:
+            W += 1j * self.A.source.random(k, distribution='normal')
+
+        offset = len(self._Q[0])
+        self._Q[0].append(self.A.apply(W))
+        gram_schmidt(self._Q[0], self.range_product, offset=offset, copy=False)
+
+        for i in range(self.subspace_iterations):
+            i = 2*i + 1
+
+            k = len(self._Q[i-1]) - offset  # check if GS removed vectors
```

I think we need to check here for `k == 0`. If that happens, then `self._Q[i-1][-k:]` would select all vectors instead of none. Not sure if `k == 0` should raise an error. Probably?

---

I actually can't remember now if I added this out of necessity or caution. I cannot construct an example where `k==0` so I wonder whether it would even be ok to completely remove this check. What do you think?

---

Aha, I found the origin https://github.com/pymor/pymor/pull/1753#discussion_r982529559.